File: Visual Studio Product Team/Official Visual Studio 2008 C# Samples/[C#]-Official Visual Studio 2008 C# Samples/Reflector/C#/Reflector.cs
//Copyright (C) Microsoft Corporation.  All rights reserved.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Text;
using System.Xml;
using System.Xml.Linq;

// The Reflector component generates a HTML document outlining the public API
// for a given assembly. The task is achieved in two phases. First, a XML 
// document is emitted in memory (see Emit*()). The document captures the 
// relevant aspects of the assembly's metadata. In the second phase, a HTML
// document is extracted from the XML document (see Extract*()). Note that the 
// two phases handle similar information with different representations using 
// the same query constructs.

namespace Samples
{   
    public class Reflector
    {
        public XDocument document;
        
        public Reflector() 
        {
        }
        
        public void Reflect(string assemblyFile)
        {
            Assembly assembly = Assembly.LoadFrom(assemblyFile);
            document = new XDocument(EmitAssembly(assembly));
        }

        public void Transform(XmlWriter writer) 
        {
            if (document == null) return;
            XElement assembly = document.Element("assembly");
            XDocument transform = new XDocument(ExtractAssembly(assembly));
            transform.Save(writer);
        }

        XElement EmitAssembly(Assembly assembly) 
        {
            return new XElement("assembly",
                        new XAttribute("name", assembly.ManifestModule.Name),
                        from type in assembly.GetTypes()
                        where GetVisible(type)
                        group type by GetNamespace(type) into g
                        orderby g.Key
                        select EmitNamespace(g.Key, g));
        }
        
        XElement EmitNamespace(string ns, IEnumerable<Type> types)
        {
            return new XElement("namespace",
                        new XAttribute("name", ns),
                        from type in types
                        orderby type.Name
                        select EmitType(type));
        }
        
        XElement EmitType(Type type)
        {
            return new XElement(type.IsEnum ? "enum" : 
                                type.IsValueType ? "struct" :
                                type.IsInterface ? "interface" :
                                "class",
                        new XAttribute("name", type.Name),
                        !type.IsGenericTypeDefinition ? null : 
                        EmitGenericArguments(type.GetGenericArguments()),
                        EmitModifiers(type),
                        EmitExtends(type.BaseType),
                        EmitImplements(type.GetInterfaces()),
                        EmitDeclaringType(type.DeclaringType),
                        EmitNestedTypes(type.GetNestedTypes()),
                        EmitMethods(type.GetConstructors()),
                        EmitProperties(type.GetProperties()),
                        EmitMethods(type.GetMethods()));
        }

        IEnumerable<XElement> EmitGenericArguments(IEnumerable<Type> args)
        {
            return from arg in args
                   select new XElement("genericArgument", EmitReference(arg));
        }
        
        static XElement EmitModifiers(Type type)
        {
            StringBuilder builder = new StringBuilder();
            if (type.IsPublic) builder.Append("public");
            else if (type.IsNestedPublic) builder.Append("public");
            else if (type.IsNestedFamily) builder.Append("protected");
            else if (type.IsNestedFamANDAssem) builder.Append("protected internal");
            if (type.IsSealed) builder.Append(" sealed");
            if (type.IsAbstract) builder.Append(" abstract");
            return new XElement("modifiers", builder.ToString());
        }
        
        XElement EmitExtends(Type baseType) 
        {
            if (baseType == null || baseType == typeof(Object) || baseType == typeof(ValueType) || baseType == typeof(Enum)) return null;
            return new XElement("extends", EmitReference(baseType));
        }
        
        IEnumerable<XElement> EmitImplements(IEnumerable<Type> ifaces)
        {
            return from iface in ifaces
                   select new XElement("implements", EmitReference(iface));
        }
        
        XElement EmitDeclaringType(Type declaringType)
        {
            if (declaringType == null) return null;
            return new XElement("declaringType", EmitReference(declaringType));
        }
        
        IEnumerable<XElement> EmitNestedTypes(IEnumerable<Type> ntypes)
        {
            return from ntype in ntypes
                   where GetVisible(ntype)
                   select EmitType(ntype);
        }
        
        IEnumerable<XElement> EmitMethods(IEnumerable<MethodBase> metds)
        {
            return from metd in metds
                   where GetVisible(metd)
                   select new XElement("method",                                
                                new XAttribute("name", metd.Name),
                                !metd.IsGenericMethodDefinition ? null :
                                EmitGenericArguments(metd.GetGenericArguments()),
                                EmitModifiers(metd),
                                EmitReturnType(metd),
                                !metd.IsDefined(typeof(ExtensionAttribute), true) ? null :
                                EmitExtension(metd),
                                EmitParameters(metd.GetParameters()));                                
        }
        
        IEnumerable<XElement> EmitProperties(IEnumerable<PropertyInfo> props)
        {
            return from prop in props
                   where GetVisible(prop.GetGetMethod()) ||
                         GetVisible(prop.GetSetMethod())
                   select new XElement("property",                            
                                new XAttribute("name", prop.Name),
                                new XElement("propertyType", EmitReference(prop.PropertyType)));
        }

        IEnumerable<object> EmitReference(Type type)
        {
            if (!type.IsGenericType) 
            {
                return new object[] { new XAttribute("name", type.Name),
                                      new XAttribute("namespace", GetNamespace(type)) };            
            }
            else
            {
                return new object[] { new XAttribute("name", type.Name),
                                      new XAttribute("namespace", GetNamespace(type)),
                                      EmitGenericArguments(type.GetGenericArguments()) };
            }
        }

        static XElement EmitModifiers(MethodBase metd)
        {
            StringBuilder builder = new StringBuilder();            
            if (metd.IsPublic) builder.Append("public");
            else if (metd.IsFamily) builder.Append("protected");
            else if (metd.IsFamilyAndAssembly) builder.Append("protected internal");
            if (metd.IsAbstract) builder.Append(" abstract");
            if (metd.IsStatic) builder.Append(" static");
            if (metd.IsVirtual) builder.Append(" virtual");
            return new XElement("modifiers", builder.ToString());
        }

        XElement EmitReturnType(MethodBase metd)
        {  
            MethodInfo metdInfo = metd as MethodInfo;
            if (metdInfo == null) return null;
            return new XElement("returnType", EmitReference(metdInfo.ReturnType));
        }

        static XElement EmitExtension(MethodBase metd)
        {
            return new XElement("extension");
        }
        
        IEnumerable<XElement> EmitParameters(IEnumerable<ParameterInfo> parms)
        {
            return from parm in parms
                   select new XElement("parameter",
                                new XAttribute("name", parm.Name),
                                new XElement("parameterType", EmitReference(parm.ParameterType)));
        }                                                
        
        static string GetNamespace(Type type)
        {
            string ns = type.Namespace;
            return ns != null ? ns : string.Empty;
        }       

        static bool GetVisible(Type type)
        {
            return type.IsPublic || type.IsNestedPublic || type.IsNestedFamily || type.IsNestedFamANDAssem;
        }
       
        static bool GetVisible(MethodBase metd)
        {
            return metd != null && (metd.IsPublic || metd.IsFamily || metd.IsFamilyAndAssembly);
        }
        
        XElement ExtractAssembly(XElement assembly)
        {
            return new XElement("html",
                        new XElement("head",
                            new XElement("title", ExtractName(assembly))),
                        new XElement("body",
                            new XElement("div", 
                                new XElement("h1", "Assembly: ",ExtractName(assembly)),
                                from ns in assembly.Elements("namespace") 
                                select ExtractNamespace(ns))));                                    
        }

        XElement ExtractNamespace(XElement ns)
        {
            return new XElement("div",
                        new XElement("h2", "Namespace: ", ExtractName(ns)),
                        from name in new string[] { "class", "interface", "struct", "enum" }
                        where ns.Elements(name).Any()
                        select from type in ns.Elements(name)
                               where !type.Elements("declaringType").Any()
                               select ExtractType(type));
        }

        XElement ExtractType(XElement type) 
        {
            return new XElement("div",
                        new XElement("h3",
                            ExtractModifiers(type) + " ",
                            type.Name + " ",
                            ExtractReference(type),                   
                            ExtractInherits(type)),
                        ExtractConstructors(type),
                        ExtractProperties(type),
                        ExtractOperators(type),
                        ExtractMethods(type));
        }

        static string ExtractModifiers(XElement element)
        {
            return element.Element("modifiers").Value;
        }

        static string ExtractName(XElement element)
        {
            string name = element.Attribute("name").Value;
            int i = name.LastIndexOf("`", StringComparison.Ordinal);
            if (i > 0) name = name.Substring(0, i); // fix generic name
            return name;
        }                        

        string ExtractGenericArguments(XElement element)
        {
            if (!element.Elements("genericArgument").Any()) return string.Empty;
            StringBuilder builder = new StringBuilder("<");
            foreach (XElement genericArgument in element.Elements("genericArgument"))
            {
                if (builder.Length != 1) builder.Append(", ");
                builder.Append(ExtractReference(genericArgument));
            }
            builder.Append(">");
            return builder.ToString();
        }

        string ExtractReference(XElement element)
        {
            return ExtractName(element) + ExtractGenericArguments(element);
        }
                
        string ExtractInherits(XElement type)
        {
            if (!type.Elements("extends").Concat(type.Elements("implements")).Any()) return string.Empty;
            StringBuilder builder = new StringBuilder();
            foreach (XElement inherits in type.Elements("extends").Concat(type.Elements("implements")))
            {
                if (builder.Length == 0) builder.Append(" : ");
                else builder.Append(", ");
                builder.Append(ExtractReference(inherits));
            }
            return builder.ToString();
        }

        XElement ExtractConstructors(XElement type)
        {
            var ctors = from ctor in type.Elements("method")
                        where ExtractName(ctor) == ".ctor"
                        select new XElement("li",
                                    ExtractModifiers(ctor) + " ",
                                    ExtractName(type),
                                    ExtractParameters(ctor));
            if (!ctors.Any()) return null;                                    
            return new XElement("div", 
                        new XElement("h4", "Constructors: "),
                        new XElement("ul", ctors));
        }

        XElement ExtractProperties(XElement type)
        {
            var props = from prop in type.Elements("property")
                        let propName = ExtractName(prop)
                        let getter = "get_" + propName
                        let setter = "set_" + propName
                        select new XElement("ul",
                                    propName,
                                    from metd in type.Elements("method")
                                    let metdName = ExtractName(metd)
                                    where metdName == getter ||
                                          metdName == setter
                                    select ExtractMethod(metd));
            if (!props.Any()) return null;                                    
            return new XElement("div", 
                        new XElement("h4", "Properties: "), 
                        props);
        }

        XElement ExtractOperators(XElement type)
        {
            var ops = from op in type.Elements("method")
                      let name = ExtractName(op)
                      where name.StartsWith("op_", StringComparison.Ordinal)
                      select new XElement("ul",
                                    name.Substring("op_".Length),
                                    ExtractMethod(op));
            if (!ops.Any()) return null;                                    
            return new XElement("div", 
                        new XElement("h4", "Operators: "),
                        ops);                      
        }
                
        XElement ExtractMethods(XElement type)
        {
            var metds = from metd in type.Elements("method")
                        let name = ExtractName(metd)
                        where name != ".ctor" &&
                              !type.Elements("property").
                               Where(prop => name == "get_" + ExtractName(prop) ||
                                             name == "set_" + ExtractName(prop)).
                               Any() &&
                              !name.StartsWith("op_") 
                        select ExtractMethod(metd);
            if (!metds.Any()) return null;
            return new XElement("div", 
                        new XElement("h4", "Methods: "), 
                        new XElement("ul", metds));
        }

        XElement ExtractMethod(XElement metd)
        {
            return new XElement("li",
                        ExtractModifiers(metd) + " ",
                        ExtractReference(metd.Element("returnType")) + " ",
                        ExtractReference(metd),
                        ExtractParameters(metd));                         
        }
        
        string ExtractParameters(XElement metd)
        {
            StringBuilder builder = new StringBuilder("(");
            foreach (XElement parm in metd.Elements("parameter"))
            {
                if (builder.Length == 1)
                {
                    if (metd.Element("extension") != null) builder.Append("this ");
                }
                else
                {
                    builder.Append(", ");
                }
                builder.Append(ExtractReference(parm.Element("parameterType")));
                builder.Append(" ");
                builder.Append(ExtractName(parm));
            }
            builder.Append(")");
            return builder.ToString();
        }
    }
}

File: Visual Studio Product Team/Official Visual Studio 2008 VB.NET Samples/[VB]-Official Visual Studio 2008 VB.NET Samples/Reflector Sample (Visual Basic)/VB/Reflector/Reflector.vb
'Copyright (C) Microsoft Corporation.  All rights reserved.

Imports System.Reflection
Imports System.Runtime.CompilerServices
Imports System.Text
Imports System.Xml

' The Reflector component generates an HTML document outlining the public API
' for a given assembly. The task is achieved in two phases. First, an XML 
' document is emitted in memory (see Emit*()). The document captures the 
' relevant aspects of the assembly's metadata. In the second phase, an HTML
' document is extracted from the XML document (see Extract*()). Note that the 
' two phases handle similar information with different representations using 
' the same query constructs.


Public Class Reflector
    Public document As XDocument

    Public Sub Reflect(ByVal assemblyFile As String)
        Dim Assembly As Assembly = Assembly.LoadFrom(assemblyFile)
        document = New XDocument(EmitAssembly(Assembly))
    End Sub

    Public Sub Transform(ByVal writer As XmlWriter)
        If document Is Nothing Then Return
        Dim assembly As XElement = document.Element("assembly")
        Dim transform As New XDocument(ExtractAssembly(assembly))
        transform.Save(writer)
    End Sub

    Public Function EmitAssembly(ByVal assembly As Assembly) As XElement
        Return <assembly name=<%= assembly.ManifestModule.Name %>>
                   <%= From type In assembly.GetTypes() _
                       Where GetVisible(type) _
                       Group type By GetNamespace = GetNamespace(type) Into g = Group _
                       Order By GetNamespace _
                       Select EmitNamespace(GetNamespace, g) %>
               </assembly>
    End Function

    Public Function EmitNamespace(ByVal ns As String, ByVal types As IEnumerable(Of Type)) As XElement
        Return <namespace name=<%= ns %>>
                   <%= From type In types _
                       Order By type.Name _
                       Select EmitType(type) %>
               </namespace>
    End Function

    Public Function EmitType(ByVal type As Type) As XElement
        Return <<%= If(type.IsEnum, "enum", If(type.IsValueType, "struct", If(type.IsInterface, "interface", "class"))) %> name=<%= type.Name %>>
                   <%= If(Not type.IsGenericTypeDefinition, Nothing, EmitGenericArguments(type.GetGenericArguments())) %>
                   <%= EmitModifiers(type) %>
                   <%= EmitExtends(type.BaseType) %>
                   <%= EmitImplements(type.GetInterfaces()) %>
                   <%= EmitDeclaringType(type.DeclaringType) %>
                   <%= EmitNestedTypes(type.GetNestedTypes()) %>
                   <%= EmitMethods(type.GetConstructors()) %>
                   <%= EmitProperties(type.GetProperties()) %>
                   <%= EmitMethods(type.GetMethods()) %>
               </>
    End Function

    Public Function EmitGenericArguments(ByVal args As IEnumerable(Of Type)) As IEnumerable(Of XElement)
        Return From arg In args _
                Select <genericArgument>
                           <%= EmitReference(arg) %>
                       </genericArgument>
    End Function

    Public Function EmitModifiers(ByVal type As Type) As XElement
        Dim builder As StringBuilder = New StringBuilder()
        If type.IsPublic Then
            builder.Append("public")
        ElseIf type.IsNestedPublic Then
            builder.Append("public")
        ElseIf type.IsNestedFamily Then
            builder.Append("protected")
        ElseIf type.IsNestedFamANDAssem Then
            builder.Append("protected internal")
        End If
        If type.IsSealed Then builder.Append(" sealed")
        If type.IsAbstract Then builder.Append(" abstract")

        Return <modifiers>
                   <%= builder %>
               </modifiers>
    End Function

    Public Function EmitExtends(ByVal baseType As Type) As XElement
        If baseType Is Nothing Or baseType Is GetType(Object) Or baseType Is GetType(ValueType) Or baseType Is GetType(System.Enum) Then Return Nothing
        Return <extends>
                   <%= EmitReference(baseType) %>
               </extends>
    End Function

    Public Function EmitImplements(ByVal ifaces As IEnumerable(Of Type)) As IEnumerable(Of XElement)
        Return From iface In ifaces _
               Select <implements>
                          <%= EmitReference(iface) %>
                      </implements>
    End Function

    Public Function EmitDeclaringType(ByVal declaringType As Type) As XElement
        If declaringType Is Nothing Then Return Nothing
        Return <declaringType>
                   <%= EmitReference(declaringType) %>
               </declaringType>
    End Function

    Public Function EmitNestedTypes(ByVal ntypes As IEnumerable(Of Type)) As IEnumerable(Of XElement)
        Return From ntype In ntypes _
                Where GetVisible(ntype) _
                Select EmitType(ntype)
    End Function

    Public Function EmitMethods(ByVal metds As IEnumerable(Of MethodBase)) As IEnumerable(Of XElement)
        Return From metd In metds _
        Where GetVisible(metd) _
        Select <method name=<%= metd.Name %>>
                   <%= If(Not metd.IsGenericMethodDefinition, Nothing, EmitGenericArguments(metd.GetGenericArguments())) %>
                   <%= EmitModifiers(metd) %>
                   <%= EmitReturnType(metd) %>
                   <%= If(Not metd.IsDefined(GetType(ExtensionAttribute), True), Nothing, EmitExtension(metd)) %>
                   <%= EmitParameters(metd.GetParameters()) %>
               </method>
    End Function

    Public Function EmitProperties(ByVal props As IEnumerable(Of PropertyInfo)) As IEnumerable(Of XElement)

        Return From prop In props _
                    Where GetVisible(prop.GetGetMethod()) Or _
                    GetVisible(prop.GetSetMethod()) _
                    Select <property name=<%= prop.Name %>>
                               <propertyType><%= EmitReference(prop.PropertyType) %>
                               </propertyType>
                           </property>

    End Function

    Public Function EmitReference(ByVal type As Type) As IEnumerable(Of Object)

        If (Not type.IsGenericType) Then
            Return New Object() {New XAttribute("name", type.Name), _
                                 New XAttribute("namespace", GetNamespace(type))}
        Else
            Return New Object() {New XAttribute("name", type.Name), _
                                 New XAttribute("namespace", GetNamespace(type)), _
                                 EmitGenericArguments(type.GetGenericArguments())}
        End If
    End Function

    Public Function EmitModifiers(ByVal metd As MethodBase) As XElement
        Dim builder As StringBuilder = New StringBuilder()
        If metd.IsPublic Then
            builder.Append("public")
        ElseIf metd.IsFamily Then
            builder.Append("protected")
        ElseIf metd.IsFamilyAndAssembly Then
            builder.Append("protected internal")
        End If

        If metd.IsAbstract Then builder.Append(" abstract")
        If metd.IsStatic Then builder.Append(" static")
        If metd.IsVirtual Then builder.Append(" virtual")

        Return New XElement("modifiers", builder.ToString())
    End Function

    Public Function EmitReturnType(ByVal metd As MethodBase) As XElement
        Dim metdInfo As MethodInfo = TryCast(metd, MethodInfo)
        If metdInfo Is Nothing Then Return Nothing
        Return <returnType>
                   <%= EmitReference(metdInfo.ReturnType) %>
               </returnType>
    End Function

    Public Function EmitExtension(ByVal metd As MethodBase) As XElement
        Return <extension>
               </extension>
    End Function

    Public Function EmitParameters(ByVal parms As IEnumerable(Of ParameterInfo)) As IEnumerable(Of XElement)
        Return From parm In parms _
               Select <parameter name=<%= parm.Name %>>
                          <parameterType>
                              <%= EmitReference(parm.ParameterType) %>
                          </parameterType>
                      </parameter>
    End Function

    Public Shared Function GetNamespace(ByVal type As Type) As String
        Dim ns As String = type.Namespace
        Return If(ns IsNot Nothing, ns, String.Empty)
    End Function

    Public Shared Function GetVisible(ByVal type As Type) As Boolean
        Return type.IsPublic Or type.IsNestedPublic Or type.IsNestedFamily Or type.IsNestedFamANDAssem
    End Function

    Public Shared Function GetVisible(ByVal metd As MethodBase) As Boolean
        Return metd IsNot Nothing AndAlso (metd.IsPublic Or metd.IsFamily Or metd.IsFamilyAndAssembly)
    End Function

    Public Function ExtractAssembly(ByVal assembly As XElement) As XElement
        Return <html>
                   <head>
                       <title>
                           <%= ExtractName(assembly) %>
                       </title>
                   </head>
                   <body>
                       <div>
                           <h1>Assembly: <%= ExtractName(assembly) %></h1>
                           <%= From ns In assembly.Elements("namespace") _
                               Select ExtractNamespace(ns) %>
                       </div>
                   </body>
               </html>
    End Function

    Public Function ExtractNamespace(ByVal ns As XElement) As XElement
        Return <div>
                   <h2>Namespace: <%= ExtractName(ns) %></h2>
                   <%= From name In New String() {"class", "interface", "struct", "enum"} _
                       Where (ns.Elements(name).Any()) _
                       Select From type In ns.Elements(name) _
                       Where Not type.Elements("declaringType").Any() _
                       Select ExtractType(type) %>
               </div>
    End Function

    Public Function ExtractType(ByVal type As XElement) As XElement
        Return <div>
                   <h3><%= ExtractModifiers(type) + " " %>
                       <%= type.Name.ToString & " " %>
                       <%= ExtractReference(type) %>
                       <%= ExtractInherits(type) %></h3>
                   <%= ExtractConstructors(type) %>
                   <%= ExtractProperties(type) %>
                   <%= ExtractOperators(type) %>
                   <%= ExtractMethods(type) %>
               </div>
    End Function

    Public Function ExtractModifiers(ByVal element As XElement) As String
        Return element.Element("modifiers").Value
    End Function

    Public Function ExtractName(ByVal element As XElement) As String
        Dim name As String = element.Attribute("name").Value
        Dim i As Integer = name.LastIndexOf("`")
        If i > 0 Then name = name.Substring(0, i) ' fix generic name
        Return name
    End Function

    Public Function ExtractGenericArguments(ByVal element As XElement) As String
        If Not element.Elements("genericArgument").Any() Then Return String.Empty
        Dim builder As StringBuilder = New StringBuilder("<")
        For Each genericArgument As XElement In element.Elements("genericArgument")
            If builder.Length <> 1 Then builder.Append(", ")
            builder.Append(ExtractReference(genericArgument))
        Next
        builder.Append(">")
        Return builder.ToString()
    End Function

    Public Function ExtractReference(ByVal element As XElement) As String
        Return ExtractName(element) + ExtractGenericArguments(element)
    End Function

    Public Function ExtractInherits(ByVal type As XElement) As String
        If Not type.Elements("extends").Concat(type.Elements("implements")).Any() Then Return String.Empty
        Dim builder As New StringBuilder()
        For Each [inherits] As XElement In type.Elements("extends").Concat(type.Elements("implements"))
            If builder.Length = 0 Then
                builder.Append(" : ")
            Else
                builder.Append(", ")
            End If
            builder.Append(ExtractReference([inherits]))
        Next
        Return builder.ToString()
    End Function

    Public Function ExtractConstructors(ByVal type As XElement) As XElement
        Dim ctors = From ctor In type.Elements("method") _
                    Where ExtractName(ctor) = ".ctor" _
                    Select <li>
                               <%= ExtractModifiers(ctor) + " " %>
                               <%= ExtractName(type) %>
                               <%= ExtractParameters(ctor) %>
                           </li>
        If Not ctors.Any() Then Return Nothing
        Return <div>
                   <h4>Constructors: </h4>
                   <ul><%= ctors %></ul>
               </div>
    End Function

    Public Function ExtractProperties(ByVal type As XElement) As XElement
        Dim props = From prop In type.Elements("property") _
        Let propName = ExtractName(prop) _
        Let getter = "get_" + propName _
        Let setter = "set_" + propName _
                        Select <ul>
                                   <%= propName %>
                                   <%= From metd In type.Elements("method") _
                                       Let metdName = ExtractName(metd) _
                                       Where metdName = getter Or metdName = setter _
                                       Select ExtractMethod(metd) %>
                               </ul>
        If Not props.Any() Then Return Nothing
        Return <div><h4>Properties: </h4><%= props %></div>
    End Function

    Public Function ExtractOperators(ByVal type As XElement) As XElement
        Dim ops = From op In type.Elements("method") _
        Let name = ExtractName(op) _
        Where name.StartsWith("op_") _
                      Select <ul>
                                 <%= name.Substring("op_".Length) %>
                                 <%= ExtractMethod(op) %>
                             </ul>
        If Not ops.Any() Then Return Nothing
        Return <div><h4>Operators: </h4><%= ops %></div>
    End Function

    Public Function ExtractMethods(ByVal type As XElement) As XElement
        Dim metds = From metd In type.Elements("method") _
                    Let name = ExtractName(metd) _
                    Where name <> ".ctor" And _
                          Not type.Elements("property").Where(Function(prop) name = "get_" & ExtractName(prop) Or _
                                                                             name = "set_" + ExtractName(prop)).Any() And Not name.StartsWith("op_") _
                    Select ExtractMethod(metd)

        If Not metds.Any() Then Return Nothing
        Return New XElement("div", New XElement("h4", "Methods: "), New XElement("ul", metds))
    End Function

    Public Function ExtractMethod(ByVal metd As XElement) As XElement
        Return <li>
                   <%= ExtractModifiers(metd) + " " %>
                   <%= ExtractReference(metd.Element("returnType")) + " " %>
                   <%= ExtractReference(metd) %>
                   <%= ExtractParameters(metd) %>
               </li>
    End Function

    Public Function ExtractParameters(ByVal metd As XElement) As String

        Dim builder As StringBuilder = New StringBuilder("(")
        For Each parm As XElement In metd.Elements("parameter")
            If builder.Length = 1 Then
                If metd.Element("extension") IsNot Nothing Then builder.Append("this ")
            Else
                builder.Append(", ")
            End If
            builder.Append(ExtractReference(parm.Element("parameterType")))
            builder.Append(" ")
            builder.Append(ExtractName(parm))
        Next
        builder.Append(")")
        Return builder.ToString()
    End Function
End Class